lib/Ordered.fram

```diff
@@ -0,0 +1,16 @@
+(* This file is part of DBL, released under MIT license.
+ * See LICENSE for details.
+*)
+
+pub data Comparable = Eq| Noteq
+
+pub data Ordered = 
+| Less
+| Equal
+| Greater
+
+pub data Interval Value = Inclusion of Value | Exclusion of Value
```

Comment:
Where it should be than in new module, because there can be a lot of code that uses this type without using sets?

Reply:
I still think it should be just in module Set. And if it is used Map, type so simple like this can be duplicated to module Map.
If there should be more code using this type it can be referenced with 
```
import Set
...
Set.Interval
```

Reply:
I strongly disagree with your statement. It's due to the reason that interval can be used in more places than just in sets or maps. The main reason is that duplicating code is a bad habit. It only leads to different implementations and versions of the same data type. This was already pointed out during the conversation about red black trees.

Reply:
While code duplication is a bad habit, it applies to complicated parts of code that will be maintained, have possibly multiple bugs to them, for which implementation of red-black trees is a perfect example. That isn't the case here. This is simple data type. Basically it's $X + X$, which is isomorphic with many other possible use cases. And it doesn't have too much functionality attached to it, so there isn't anything to maintain.

The other bad habit that is exemplified here is tight coupling. It's bad because it couples parts of code base that don't have too much to do with each other. Here it's harmful because of generality of $X + X$. You names it `Interval`, which is very suggestive, as are names of constructors you've chosen. I'm sure there are many usages for $X + X$, where those names don't mean anything, but user might be tempted to use them because they are already implemented. That will lead to confusing code that requires value of type `Interval`, where it doesn't mean anything in a best case scenario, or worse  it could be counterintuitive.

Now let's say that in the future, for some reason we want to change how this part works. You are stuck. You can't change just this type for what you want it to be, because there are so many other parts of code base that (maybe not entirely correctly) use this type.

And finally it's bad at doing what you intend it to do. You need to import module to use one function of a different module? In my opinion this whole `Ordered` is a bad idea. Let's say you want to create set of Ints. Now instead of just writing `Map.make {Key=Int}` or something like that, you have to write function that compares two Ints and says if they are different or equal... In my opinion type Ordered should be defined in Prelude, maybe as a module. While Maps and Sets should take in explicit type and expect that it implements something like method compare, because it much easier to write `Map.make {Key=Int}` than to search for where is the comparator defined, or even define it yourself.
But also this is a valid use case so it wouldn't hurt to have another function like `makeWithComparator`.

Reply:
I can change the name from Ordered to Relationship and then you don't have problem with having in the same file Comparable, Ordered and Interval. 
Secondly you saying that now there is no much functionality but it might change in the future. Secondly I will not stop someone from shooting his foot if that person what to use a type which have specifically and intentionally named constructors. 
Thirdly the construction of Map and Set was discussed a long time ago, where the type Ordered was approved. And if you looked on other implementation of sets and maps, you would find that most of them have some type defined as ordered. For example in Ocaml we have 
`
module type OrderedType = sig 
 type t
 val compare : t -> t -> int
end
`

Reply:
> I can change the name from Ordered to Relationship and then you don't have problem with having in the same file Comparable, Ordered and Interval.

Type name isn't the problem here, problem is that it should be defined in Prelude and somehow integrated, so users don't have to implement compare for ints every time they want to use map of ints. 

> Secondly you saying that now there is no much functionality but it might change in the future.

There is only so much you can do with so simple type. I can see extension to general idea of interval, but not for how it's defined here. And again if we were to extend it we would have to do it keeping in mind everyone also using this type, even for how it's intended.

As for how it's implemented elsewhere. Interface of such things is dependent on language it's written for. Ocaml uses module types, because that's the best they can do. And it allows them to quickly create Maps of simple types, just passing in the module with appropriate functionality. We can do better thanks to for example methods. On the other hand while we could do functors exactly like they are done in Ocaml, but it's sub optimal.

Reply:
I agree (mostly) with @Foxinio. I see no problem with duplicating such simple definitions like `Interval`: it is just a type without any functionality, so duplicating it doesn't have any bad consequences. Moreover, this type is not an interval *per se*, and have the only one application, so it should stay (private) in this module.

The situation with `Ordered` (or maybe `Order` or just `Ord`) is different. It has many application, because it defines order on a type that may be used by many data structures. I should be definitely defined in `Prelude`. I would also change the names of constructors to `Lt`, `Eq`, and `Gt`, because full world are very long.

> As for how it's implemented elsewhere. Interface of such things is dependent on language it's written for. Ocaml uses module types, because that's the best they can do. And it allows them to quickly create Maps of simple types, just passing in the module with appropriate functionality. We can do better thanks to for example methods. On the other hand while we could do functors exactly like they are done in Ocaml, but it's sub optimal.

That's the point! We should not copy idioms from other languages, but try to figure out what will work best in our case, because our language is different.